### PR TITLE
Examples/benchmarking

### PR DESCRIPTION
**Pull request by tdiethe**
*Description of changes:*

This is some benchmarking of Bayesian Neural Networks (meanfield VI) against a non-Bayesian NN. Hopefully this could provide a useful starting point for further analysis (e.g. different kinds of BNN). 

The script `examples/benchmarking/bnn_classification_benchmark.py` runs through several datasets (MNIST, FashionMNIST, CIFAR10, CIFAR100), with 3 different NN architectures. Several metrics are computed (Accuracy, MSE (=Brier score), Log loss). Some "sensible" defaults are set for the hyperparameters - no HP tuning is performed. Results are stored in the `results.txt` file as a list of json strings.

Also added a notebook in the notebooks directory for exploring the results. This outputs figures to the directory `examples/benchmarking/figs` (figures also included). 

Changes to MXFusion core files:
- `mxfusion/components/functions/mxfusion_gluon_function.py`: Made the exception more helpful
- `mxfusion/inference/batch_loop.py`: added a callback for custom status messages
- `mxfusion/inference/grad_based_inference.py`: added `GradIteratorBasedInference` - a version of `GradBasedInference` that operates on a data loader
- `mxfusion/inference/minibatch_loop.py`: fixed bug that stopped it working on GPUs; added callback for custom status messages

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by tdiethe** [comment]
@meissnereric can you have a look at the failing tests? Don't think this was happening before.

**Comment by meissnereric** [comment]
I think this was happening before, I remember seeing it.

The reason is that you're using Python 3.6 only string formatting in places. this "(f"Context device id {ctx.device_id} outside range of list {ctx_list} or None")" style isn't supported in 3.4/3.5, use the classic "blah".format() style.
